Refactor the xarch instructions to be in a InstructionInfo array

tannergooding opened the pull request:
This refactors the xarch `instruction` tables to be defined in a single array of `InstructionInfo` structures.

tannergooding commented:
FYI. @CarolEidt, @fiigii 

This is the first PR of the refactoring and is the "minimal" changes needed. The general "layout" of the `InstructionInfo` struct is similar to the `HWIntrinsicInfo` struct.

The second PR will rewrite `instrlistxarch.h`. This will include moving the encodings into a `size_t[]` and having a `flags` field that will contain various data (including things like `IsDstDstSrc`, `IsDstSrcSrc`, etc)

tannergooding commented:
This is a small part of https://github.com/dotnet/coreclr/issues/14523

tannergooding commented:
test this please

tannergooding commented:
The second PR is going to be this branch: https://github.com/tannergooding/coreclr/tree/instr-refac2

The two primary commits add flags for DstSrcSrc (https://github.com/tannergooding/coreclr/commit/2b9580036fadc20cdb061b3cc6d629bf3a73c1c7) and  DstDstSrc (https://github.com/tannergooding/coreclr/commit/c8b8b8f9eb756b7f953ccf51b12e1ab1988140d2)

There are a at least a few other checks that could be table driven "easily" as well (such as the "Takes VEX.W Check").

This also makes it significantly easier to do other refactorings (such as ordering the instructions better, having a Non-VEX vs VEX operand count field, etc).

mikedn commented:
> The two primary commits add flags for DstSrcSrc and DstDstSrc 

Why are these changes needed to add a couple of flags? The existing "fp" field can already be used for that.

> such as ordering the instructions better

Why is that needed?

tannergooding commented:
> Why are these changes needed to add a couple of flags? The existing "fp" field can already be used for that.

The current setup was designed with the legacy backend and is difficult to work with:
* It is non-obvious which fields are used and which aren't
* It is non-obvious all of the items that touch/consume the instruction list
  * Today we have 5-6 separate lookup arrays
* The definitions have to be ordered in a particular way, regardless of whether that makes the code easier to browse
  * INST5, then INST4, ... then INST0
* There are more than 8 flags that will need to be eventually added
  * 5 already, plus DstSrcImm, TakesVexPrefix, TakesRexWPrefix, supports 2-byte VEX encoding, etc...
* It is not easy to add support for new encodings/etc
  * This has lead to several cases where we just reuse an existing encoding slot and have special checks or handling for it
    At the very least, we can probably use a flag to control this better since we don't want to arbitrarily grow the tables
* etc

> Why is that needed?

Right now the instructions are mish-mashed together with vary loose ordering.

We have some general purpose instructions, then some SIMD instructions (where we have SSE/SSE2 and some SSE3 instructions combined together, 4-byte SSE3/SSE4 combined together, AVX/AVX2/FMA combined together), then more general purpose instructions.

This makes it hard to find the appropriate instruction (generally you have to actually search the file, rather than just scrolling down to the right section and looking for it alphabetically/etc).

This also makes it very non-obvious why things or functions behave a particular way in the emitter. For example:
* `IsAVXInstruction` is really a `IsVEXEncoded` check
* `IsSSEOrAVX` is really `IsSimdInstruction`
* `IsSSE4Instruction` is really `IsSse3OrSse4Instruction`
* etc

Overall, these changes should make it significantly easier to maintain and improve the instruction information tables.

mikedn commented:
> Today we have 5-6 separate lookup arrays

So? How does that make things unobvious? 

> The definitions have to be ordered in a particular way, regardless of whether that makes the code easier to browse 

And what exactly makes this code easier to browse? Alphabetic ordering? If so, why exactly is this change the solution?

> There are more than 8 flags that will need to be eventually added 

How is the number of flags that needs to be added relevant?

> It is not easy to add support for new encodings/etc 

That's true but it's not clear how this change addresses that. Besides, my question was about adding flags. Adding flags is clearly needed but can be done without this change.

> This makes it hard to find the appropriate instruction (generally you have to actually search the file, rather than just scrolling down to the right section and looking for it alphabetically/etc).

Since when searching a text file to find something qualifies as hard?

> This also makes it very non-obvious why things or functions behave a particular way in the emitter.

It's certainly true that the emitter code is hairy. What's not clear is how changing the tables the way you propose improves things.

> Overall, these changes should make it significantly easier to maintain and improve the instruction information tables.

I'm not convinced. This change looks very much like a fancy change that's devoid of substance.

tannergooding commented:
> I'm not convinced. This change looks very much like a fancy change that's devoid of substance.

I don't think its "fancy" at all. It is basically taking the current system and restructuring it for easier maintainabilty

What we have in the table is information on instructions, so I put that in a single `InstructionInfo` struct. All data is then looked up using methods that are members of that struct. It is basic OO programming.

It, simultaneously, makes a couple of cleanups that make it easier to maintain the code:
* It is now obvious that `fp`, `rf`, and `wf` are flags (and they are now typed enum values)
* We can move all the methods that aren't actually dependent on the compiler/emitter out of the compiler/emitter (and in some cases, it may make sense to still move them out and to just pass in the emitter as a parameter).

This change can be described as:
* Take all individual 'insInfo' arrays and merge them into one structured array
* Merge the `fp`, `rf`, and `wf` parameters into a single `flags` field
  * This also makes it easier to add new flags

The second PR (which I linked, but isn't up yet) is:
* Restructure `#define INST*` to better match the new structured array layout
* Add new flags for tracking `DstDstSrc` and `DstSrcSrc`

tannergooding commented:
It sounds like the real push-back you have here is taking the individual arrays (one per "field") and making them be a single array of structures, is that correct?

If so, could you elaborate on why you think this is a bad idea.

I do agree that several of these changes could be done independently, but (IMO) restructuring this code will also make it easier to do and gives a "clearer" picture of where the related code should go and how it should be exposed.

tannergooding commented:
> Also the file renames, that one is completely unnecessary.

I did this because `instrlist` is much clearer to differentiate than `instrs` (when compared to `instr`)

> it looks more like you simply started by moving things around

Sometimes a series of simple refactorings of the existing code is much better than a more complex refactoring. I don't think that this is entirely the right shape either, but I do think that it is a step in the right direction.

It also makes it easier to extend what we have now, for the work that is going on now.

> Hint - the table order that you keep complaining about is there for a reason. I'm not sure it's worthwhile enough but it was certainly done on purpose like that.

Right, there are a few parts to the current ordering:
1. Some subsets actually need to be ordered (such as jcc and cmovcc)
  * These can also still be kept
2. Some subsets need to be generally grouped (such as the SSE/AVX ISAs)
  * These can still be kept
3. There was some attempt to "save space" in the lookup arrays (since not every instruction supports all encodings).
  * However, most of this went out the window with the SIMD instructions, as they all require INST3
    This can also still be achieved with the current table
    * We can reduce from `size_t` to just `int` by changing the 8 AMD64 instructions that require it to use a flag
    * It looks like we can change the encoding table to be 3 entries (instead of 5) by using a flag as well

mikedn commented:
> I did this because instrlist is much clearer to differentiate than instrs (when compared to instr)

With such reasoning we'd need to rewrite the entire JIT because it would be much clear...

> Sometimes a series of simple refactorings of the existing code is much better than a more complex refactoring

Sure. Provided that you have a plan to address whatever issues the emitter has and provided that the refactoring follows the plan. And having written assemblers and disassemblers in the past I can tell you that you are pretty much guaranteed to screw up one way or another without a plan.

> Some subsets need to be generally grouped (such as the SSE/AVX ISAs)

Hmm, why? I would guess that a range check such as (x >= INS_SSE_FIRST && x <= INS_SSE_LAST) might be faster than a flag check but are such checks really necessary?

> There was some attempt to "save space" in the lookup arrays (since not every instruction supports all encodings).

Indeed. I don't think that the addition of new SIMD instructions made things worse. It's not really relevant how large the table is, what's relevant is what part of the table gets used and ends up in the CPU cache. SIMD instructions being mostly grouped together don't cost anything if they're not used (which is the common case after all).

But now you're possibly making things worse by having instructions with a single encoding take up more space. A single InstructionInfo doesn't even fit in a cache line. You also include the instruction name even if that should be either ifdefed out in release build or stored separately. And didn't order data members to avoid alignment holes.

tannergooding commented:
> With such reasoning we'd need to rewrite the entire JIT because it would be much clear...

Ideally we would do this for any code which isn't clear and ideally this would be done piece by piece.

> Sure. Provided that you have a plan to address whatever issues the emitter has and provided that the refactoring follows the plan. And having written assemblers and disassemblers in the past I can tell you that you are pretty much guaranteed to screw up one way or another without a plan.

Good thing this is a simple, iterative, PR and not a more complex refactoring.

> Hmm, why? I would guess that a range check such as (x >= INS_SSE_FIRST && x <= INS_SSE_LAST) might be faster than a flag check but are such checks really necessary?

Because that is the way it is today. This PR makes it  easier to change the ordering and to use a flag, if someone thinks that is better. However, grouping ISAs together, to me, makes sense.

> what's relevant is what part of the table gets used and ends up in the CPU cache. 

And the parts of the table that will get most frequently used are:
* The ID
* The encoding
* The flags

The General Purpose registers will likely continue being primarily encoding based lookups (they don't have a lot of behavior that depends on flags today, but they do have several checks that could become flags, rather than switch or if/else checks).

> A single InstructionInfo doesn't even fit in a cache line

Not convinced this is a problem. If it is, I believe it also exists in the current layout.

Even with the current layout, you could only fit 8 sequential instructions in a single cache line and they aren't guaranteed to be looked up or emitted sequentially to eachother (there are also no comments on perf optimizations done based on the layout/ordering and there has been no feedback on prior PRs adding instructions to order these a particular way).

Additionally, given all the other memory accesses (both reads and writes) currently required to build the instruction descriptor and then emit the resulting code, the encoding lookup being in cache is probably not a significant concern.

> You also include the instruction name even if that should be either ifdefed out in release build or stored separately

This is constructive feedback of the change. :+1:

> And didn't order data members to avoid alignment holes.

Also constructive feedback. For the second set of changes (listed above), it finishes cleaning up the table to resolve most of this (except for a possible 4 byte hole between `name` and `updateMode`).

tannergooding commented:
> I take it from this answer that you don't have any plan

The current plan is to provide iterative improvement on the existing code to better enable us to:
* Remove some of the giant switch tables we have in emitxarch
* Support additional encoding formats (such as the 2-byte VEX encoding)

A larger rewrite is something entirely different and would have involved a markdown file detailing said plan. It would have been reviewed and approved outside of a PR and would likely have had input from multiple people.

> You just happen to think that this simple, iterative, PR will magically and hopefully make things better.

Not magically. This PR is the first step. A second PR will go up after this is reviewed and merged that will get rid of several of the giant switch tables (already linked). And then a third PR will go up that adds support for the 2-byte VEX encoding (being worked on). (and I do acknowledge that merging this PR may involve only merging the flags fields together, but that will be determined based on feedback from the JIT team).

mikedn commented:
> Remove some of the giant switch tables we have in emitxarch

For the zillionth billionth time - this can be done without all these changes.

> Support additional encoding formats (such as the 2-byte VEX encoding)

How does `InstructionInfo` helps with that?

CarolEidt commented on src/jit/emit.h:
```diff
@@ -2159,10 +2161,12 @@ inline bool emitter::emitIsScnsInsDsc(instrDesc* id)
 {
-#ifdef DEBUG
+#ifdef _TARGET_XARCH_
+    return InstructionInfo::lookupUpdateMode(ins);
```
I'm not a fan of the verb "lookup". IMO, getting an entry in a table is the perfect place for the index (`[]`) operator - in fact, I'm often inclined to overload that operator even when it's not actually an array.
If a specific "get" method is desired, I would use that verb, e.g. `getUpdateMode(ins)`.

tannergooding replied:
You can't actually have static index operators in C++, I've changed the name to just `get()` for the time being.

CarolEidt commented:
@dotnet/jit-contrib - I think it would be good to get additional perspectives on the evolution of instruction encodings.

tannergooding commented:
> That said, I'm not in favor of changing instrs to instrlist

Dropped.

> I'd also prefer not to use the "lookup" verb, but would be interested in other views.

Changed to be `get`

tannergooding commented:
@CarolEidt, updated according to your (and some of @mikedn's) feedback:
* Kept the `instrs.h` and `instrs{tgt}.h` naming
* Used `get` instead of `lookup`
  * We might want to change the convention in `HWIntrinsicInfo` as well?
* Made the `id` and `name` fields `DEBUG` only
* Added the `InsEncodingFmt` and `InsFlags` enums
  * Updated the struct to use these and `#define INSTRUCTION` to assign them correctly

This PR does not rewrite `instrsxarch.h` to only use `#define INSTRUCTION`. I am still planning that for the second PR since it should reduce the noise and make this one easier to review (let me know if you would prefer otherwise -- change is already done, so it isn't any additional work).

tannergooding commented:
Applied formatting patch.

CarolEidt commented:
This looks good to me.
I'd like to get another review from someone else on @dotnet/jit-contrib.
Also, I'd like to see the throughput impact, if any (on a Release build):
- A reasonable, though far from perfect, throughput metric is the pin-icount for crossgen'ing System.Private.CoreLib.dll
- Separately, running with `COMPlus_JitMemStats=1` will give you the memory usage statistics. That would be good to see as well.

BruceForstall commented on src/jit/CMakeLists.txt:
```diff
@@ -130,6 +130,7 @@ if (WIN32)
     inline.h
     inlinepolicy.h
     instr.h
+    instrxarch.h
```
instrxarch.h _and_ instrsxarch.h -- that is going to be confusing.

tannergooding replied:
This is one of the reasons I had renamed `instrsxarch.h` t `instrslistxarch.h` originally.

BruceForstall commented on src/jit/emit.h:
```diff
@@ -1327,10 +1327,12 @@ class emitter
     insUpdateModes emitInsUpdateMode(instruction ins);
     insFormat emitInsModeFormat(instruction ins, insFormat base);
 
+#if !defined(_TARGET_XARCH_)
```
This appears to be xarch-only, so should just be removed.

tannergooding replied:
Fixed.

BruceForstall commented on src/jit/instrsxarch.h:
```diff
@@ -27,24 +24,28 @@
 
-#ifndef INST1
-#error  At least INST1 must be defined before including this file.
+#ifndef INSTRUCTION
```
INSTRUCTION should be `#undef`ed at the end of this file, so includers don't need to do that.

tannergooding replied:
Fixed.

BruceForstall commented on src/jit/instrxarch.cpp:
```diff
@@ -0,0 +1,39 @@
+
+// clang-format off
+static const InstructionInfo instructionInfoArray[] = {
+#if DEBUG
```
`#if DEBUG` => `#ifdef DEBUG` (everywhere)

tannergooding replied:
Fixed.

BruceForstall commented on src/jit/instrxarch.cpp:
```diff
@@ -0,0 +1,39 @@
+        {INS_##id, nm, {mr, mi, rm, a4, rr}, um, static_cast<InsFlag>((InsFlag_FloatingPoint*fp)|(InsFlag_ReadsFlags*rf)|(InsFlag_WritesFlags*wf))},
+#else
+    #define INSTRUCTION(id, nm, fp, um, rf, wf, mr, mi, rm, a4, rr) \
+        {{mr, mi, rm, a4, rr}, um, static_cast<InsFlag>((InsFlag_FloatingPoint*fp) | (InsFlag_ReadsFlags*rf) | (InsFlag_WritesFlags*wf))},
```
So this single table is going to have a lot of BAD_CODE entries that didn't exist (and weren't needed) in the previous tables. E.g., all INST2 will waste 3 size_t entries. I'm not sure I'm worried about that, but I do generally worry about "de-optimization" unless there's a good reason for it.

tannergooding replied:
Today, there are:
* INST5: 9 definitions. 27/45 encodings are BAD_CODE
* INST4: 11 definitions. 3/44 encodings are BAD_CODE
* INST3: 390 definitions. 764/1170 encodings are BAD_CODE
* INST2: 24 definitions. 8/48 encodings are BAD_CODE
* INST1: 48 definitions. 0/48 encodings are BAD_CODE
* INST0: 35 definitions. 1/35 encodings are BAD_CODE

This change adds 1195 new BAD_CODE encodings with this change (size increase info is here: https://github.com/tannergooding)
* The cost can be made less on x64 by changing to size_t.

I believe that the overall increased readability here is better, especially with all the HWIntrinsic work that has required new instructions to be added (and with more instructions still to come)

It may be possible to reduce the cost further by separating the SIMD instructions (the vast bulk of the encodings, which are all `INST3` today, and which generally only use the `mr` and `rm` encodings) out from the general purpose instructions.

mikedn replied:
> I believe that the overall increased readability here is better,

I don't see any increase in readability. More is usually not better. Or it's subjective.

BruceForstall commented on src/jit/instrxarch.h:
```diff
@@ -0,0 +1,107 @@
+    InsEncodingFmt_count = InsEncodingFmt_none
+};
+
+enum InsFlag : unsigned int
```
Why is this `unsigned int` instead of `unsigned char`? You'll be wasting 3 bytes per entry.

tannergooding replied:
It would need to be `unsigned short` at least. There are a number of switch tables (such as `DstDstSrc`, `DstSrcSrc`, `DstSrcImm`, `TakesRex`, `TakesRexW`, `TakesVex`, etc) which should be moved to be flags.

BruceForstall commented on src/jit/instrxarch.h:
```diff
@@ -0,0 +1,107 @@
+#endif
+
+    size_t         encoding[5];
+    insUpdateModes updateMode;
```
This is wasting 3 bytes per entry, since `insUpdateModes` is an enum that defaults to `int` type. Why isn't this part of `flags`?

tannergooding replied:
It could be changed to be two flags (UpdateModeRead and UpdateModeWrite) if desired.

BruceForstall commented on src/jit/emitxarch.cpp:
```diff
@@ -1405,10 +1212,7 @@ inline bool hasCodeMR(instruction ins)
-    assert((insCodesMR[ins] != BAD_CODE));
-
-    return insCodesMR[ins];
+    return InstructionInfo::getEncoding(ins, InsEncodingFmt_mr);
```
I'm worried that with all of these functions, we're going from a simple array access, to a double array access: one to find the instruction info array element, and one to access the encoding.

It's my assumption that these functions are hot, since I believe the encoder is hot, so giving up this performance seems bad.

tannergooding replied:
Its accessing the field of a struct in a constant array. The compiler should be optimizing this to be a single lookup.

tannergooding replied:
Also, to explicitly call it out, the `encoding` field is a C-style inline array, so there is no second indirection there.

mikedn replied:
What really happens is that we'll end up with 1 or 2 extra instructions per array access to handle the `sizeof(InstructionInfo)` multiplication required for indexing. These instructions did not exist before as those parallel arrays were primitive arrays and indexing was handled entirely by addressing mode.

tannergooding commented:
@CarolEidt:

There is a dll size increase of exactly:
* 8k (8192 bytes) for x64 (0.58%)
  * I think we can cut this down to 4k by making the encoding fields `int`, rather than `size_t`. As mentioned, only 8 instructions (`imul_r8` through `imul_r15`) actually make use of `size_t` and they could use flags to indicate the Rex prefix is required instead
* 4k (4096 bytes) for x86 (0.35%)

Running JitMemStats shows no diffs. This makes sense since the table is created at compile time and placed in the DLL data section.

pin-icount (without disabling PGO):
* 19.4 million for x64 (0.10%)
* 7.97 million for x86 (0.03%)

pin-icount (with disabling PGO):
* 6.36 million for x64 (0.03%)
* 13.0 million for x86 (0.05%)

mikedn commented:
> Running JitMemStats shows no diffs. This makes sense since the table is created at compile time and placed in the DLL data section.

The only thing that's relevant in this case is the impact on CPU cache usage. The new table format is larger. And it's pretty large already and unnecessarily so. It's hard measure the impact of this change. I don't really the expect the impact to be significant but then this PR is associated with an issue tagged "jit-throughput" so I'm baffled. No reasonable technical arguments have been provided to support this change and at the same time it seems that it carries a throughput regression risk.

> 19.4 million for x64 (0.10%)

Same comment as above.

mikedn commented:
> I think we can cut this down to 4k by making the encoding fields int, rather than size_t. As mentioned, only 8 instructions (imul_r8 through imul_r15) actually make use of size_t and they could use flags to indicate the Rex prefix is required instead

So why not fix that first?

CarolEidt commented:
@tannergooding - I hate to say this, because I think this is a step in the right direction,, but I think it probably makes sense to hold off on this until we have a discussion on direction.
@BruceForstall and I were discussing the possibility of introducing this style of interface, but consider using an SoA (struct of arrays) -style approach (like what's currently used) "under the covers" rather than the proposed AoS.
I suggest we begin that discussion in #14523 where we can begin to lay out the current issues (e.g. the numerous switch statements in the emitter) and how an alternative approach might address them.

tannergooding commented:
@CarolEidt, @BruceForstall, would it be reasonable to, at the very least, change this to just modifying the `INST` definitions to have a "flags" field (rather than three separate `fp`, `rf`, and `wf` operands that are merged behind the scenes)?

I would expect we will have this field regardless of what the ultimate layout is and doing that will allow us to getting rid of the enormous switch tables in `IsDstDstSrc` and the like.

That is, simplify this to just change:
`#define INST0(id, nm, fp, um, rf, wf, mr)` to `#define INST0(id, nm, um, flags, mr)`
and then rewrite the entries in `instrsxarch.h` to match:
`INST0(l_jg   , "jg"           , 0, IUM_RD, 1, 0, 0x008F0F)` to `INST0(l_jg   , "jg"           , IUM_RD, InsFlag_ReadsFlags, 0x008F0F)`
